lib/elixir/lib/task/supervised.ex

```diff
@@ -148,4 +147,176 @@ defmodule Task.Supervised do
   defp get_reason(reason) do
     reason
   end
+
+  ## Pmap
+
+  def pmap(enumerable, acc, reducer, mfa, options, spawn) do
+    next = &Enumerable.reduce(enumerable, &1, fn x, acc -> {:suspend, [x | acc]} end)
+    max_concurrency = Keyword.get(options, :max_concurrency, System.schedulers_online)
+    timeout = Keyword.get(options, :timeout, 5000)
+    parent = self()
+
+    # Start a process responsible for translating down messages.
+    {monitor_pid, monitor_ref} = spawn_monitor(fn -> pmap_monitor(parent) end)
+    send(monitor_pid, {parent, monitor_ref})
+
+    pmap_reduce(acc, max_concurrency, 0, 0, %{}, next,
+                reducer, mfa, spawn, monitor_pid, monitor_ref, timeout)
+  end
+
+  defp pmap_reduce({:halt, acc}, _max, _spawned, _delivered, waiting, next,
+                   _reducer, _mfa, _spawn, monitor_pid, monitor_ref, _timeout) do
+    is_function(next) && next.({:halt, []})
+    pmap_close(waiting, monitor_pid, monitor_ref)
+    {:halted, acc}
+  end
+
+  defp pmap_reduce({:suspend, acc}, max, spawned, delivered, waiting, next,
+                   reducer, mfa, spawn, monitor_pid, monitor_ref, timeout) do
+    {:suspended, acc, &pmap_reduce(&1, max, spawned, delivered, waiting, next,
+                                   reducer, mfa, spawn, monitor_pid, monitor_ref, timeout)}
+  end
+
+  # All spawned, all delivered, next is done.
+  defp pmap_reduce({:cont, acc}, _max, spawned, delivered, waiting, next,
+                   _reducer, _mfa, _spawn, monitor_pid, monitor_ref, _timeout)
+       when spawned == delivered and next == :done do
+    pmap_close(waiting, monitor_pid, monitor_ref)
+    {:done, acc}
+  end
+
+  # No more tasks to spawn because max == 0 or next is done.
+  defp pmap_reduce({:cont, acc}, max, spawned, delivered, waiting, next,
+                   reducer, mfa, spawn, monitor_pid, monitor_ref, timeout)
+       when max == 0
+       when next == :done do
+    receive do
+      {{^monitor_ref, position}, value} ->
+        waiting = Map.put(waiting, position, {:ok, value})
+        pmap_deliver(acc, max + 1, spawned, delivered, waiting, next,
+                     reducer, mfa, spawn, monitor_pid, monitor_ref, timeout)
+      {:DOWN, {^monitor_ref, position}, reason} ->
+        waiting = Map.put(waiting, position, {:exit, reason})
```

Comment:
I think there is a race condition here when a later position tasks finishes before an earlier one. In this situation the `{:DOWN, _, :normal}` can overwrite the `{:ok, value}` if the earlier position task hasn't sent its result before the monitor process ends the `:DOWN`.